cas-mfa-java/src/main/java/net/unicon/cas/mfa/authentication/RegisteredServiceMfaRoleProcessor.java

```diff
@@ -0,0 +1,254 @@
+        this.authenticationMethodConfiguration = authenticationMethodConfiguration;
+        this.servicesManager = servicesManager;
+
+        translationMap = new LinkedHashMap<String, Pattern>();
+    }
+
+    /**
+     * Resolves the authn_method for a given service if it supports mfa_role and the user has the appropriate attribute.
+     * @param authentication the user authentication object
+     * @param targetService the target service being tested
+     * @return a list (usually one) mfa authn request context.
+     */
+    public List<MultiFactorAuthenticationRequestContext> resolve(@NotNull final Authentication authentication,
+                                                                 @NotNull final WebApplicationService targetService) {
+
+        String authenticationMethodAttributeName = null;
+        final List<MultiFactorAuthenticationRequestContext> list = new ArrayList<MultiFactorAuthenticationRequestContext>();
+        if ((authentication != null) && (targetService != null)) {
+            final ServiceMfaData serviceMfaData = getServicesAuthenticationData(targetService);
+
+            if (serviceMfaData == null || !serviceMfaData.isValid()) {
+                logger.debug("No specific mfa_role service attributes found");
+                return null;
+            }
+
+            logger.debug("Found mfa_role: {}", serviceMfaData);
+
+            authenticationMethodAttributeName = serviceMfaData.attributeName;
+
+            final Object mfaAttributeValueAsObject = authentication.getPrincipal().getAttributes().get(serviceMfaData.attributeName);
+            if (mfaAttributeValueAsObject != null) {
+                if (mfaAttributeValueAsObject instanceof String) {
+                    final String mfaAttributeValue = mfaAttributeValueAsObject.toString();
+                    final MultiFactorAuthenticationRequestContext ctx = getMfaRequestContext(
+                            serviceMfaData, mfaAttributeValue, targetService);
+                    if (ctx != null) {
+                        list.add(ctx);
+                    }
+                } else if (mfaAttributeValueAsObject instanceof List) {
+                    final List<String> mfaAttributeValues = (List<String>) mfaAttributeValueAsObject;
+                    for (final String mfaAttributeValue : mfaAttributeValues) {
+                        final MultiFactorAuthenticationRequestContext ctx = getMfaRequestContext(
+                                serviceMfaData, mfaAttributeValue, targetService);
+                        if (ctx != null) {
+                            list.add(ctx);
+                        }
+                    }
+                } else {
+                    logger.debug("Not MFA Attribute found.");
+                }
+            }
+        }
+
+        if (list.size() == 0) {
+            logger.debug("No multifactor authentication requests could be resolved based on [{}]. Trying generic 'authn_method' attribute",
```

Comment:
Little concerned about hardcoding `'authn_method'` in the log. The attr name can change.